image:push authentication required

```bash
cesco@cesco-desktop: ~/code/go/src/bitbucket.org/cescoferraro/api on develop [+$]
$ VERSION=0.1.2 dobi  production:push
[WARN] Missing auth config for "https://golang/v1"
1.7.1-alpine: Pulling from library/golang
Digest: sha256:03ca8be2e13e3f83af9ce3fa181f88bddd39e83f1d2714ae7153620e9f051069
Status: Image is up to date for golang:1.7.1-alpine
[image:pull base] golang Pulled
[job:run binary] go build src/backend/api.go Start
[job:run binary] go build src/backend/api.go Done
Step 1 : FROM alpine:3.4
 ---> ee4603260daa
Step 2 : RUN apk add --no-cache ca-certificates apache2-utils
 ---> Using cache
 ---> 45118353fd77
Step 3 : ADD ./api /srv/app/api
 ---> Using cache
 ---> b53fd7b4f556
Step 4 : WORKDIR /srv/app
 ---> Using cache
 ---> a4d7f6a52822
Step 5 : CMD ./api runserver
 ---> Using cache
 ---> 94d98e1b5743
Step 6 : EXPOSE 9000
 ---> Using cache
 ---> 11fddc5aa441
Successfully built 11fddc5aa441
[image:build production] cescoferraro/api Created
[image:tag production] cescoferraro/api Tagged
[WARN] Missing auth config for "https://cescoferraro/v1"
The push refers to a repository [docker.io/cescoferraro/api]
328ac6e90ae9: Preparing 
b4b945d9e06c: Preparing 
9007f5987db3: Preparing 
[ERROR] Failed to execute task "production:push": unauthorized: authentication required
cesco@cesco-desktop: ~/code/go/src/bitbucket.org/cescoferraro/api on develop [+$]
$ docker push cescoferraro/api:v0.1.2
328ac6e90ae9: Pushed 
328ac6e90ae9: Pushing [===================================>               ] 11.14 MB/15.65 MB
b4b945d9e06c: Layer already exists 
v0.1.2: digest: sha256:7f551ac2ac64ba3535eaef411c0c4fd6ba729351ee83ff1b886ad8b48f23aa36 size: 949
```

### dobi.yaml

```
meta:
    project: api
    default: production

mount=source:
    bind: .
    path: /go/src/bitbucket.org/cescoferraro/api

image=base:
    image: golang
    tags: [1.7.1-alpine]
    pull: "always"

job=binary:
    use: base
    mounts: [source]
    command: go build src/backend/api.go
    working-dir: /go/src/bitbucket.org/cescoferraro/api
    description: "Build the static binary"

image=production:
    image: cescoferraro/api
    tags: ["v{env.VERSION}"]
    dockerfile: docker/Dockerfile.prod
    depends: [binary]
    description: "Production Image"
```

--- a/.gitignore
+++ b/.gitignore
@@ -6,3 +6,4 @@ docs/gen
 docs/build
 docs/doctrees
 tasks/client/mock_iface.go
+.glide

--- a/cmd/dobi.go
+++ b/cmd/dobi.go
@@ -37,12 +37,12 @@ func NewRootCommand() *cobra.Command {
 	var opts dobiOptions
 
 	cmd := &cobra.Command{
-		Use:                   "dobi [flags] RESOURCE[:ACTION] [RESOURCE[:ACTION]...]",
-		Short:                 "A build automation tool for Docker applications",
-		SilenceUsage:          true,
-		SilenceErrors:         true,
-		TraverseChildCommands: true,
-		Args: cobra.ArbitraryArgs,
+		Use:              "dobi [flags] RESOURCE[:ACTION] [RESOURCE[:ACTION]...]",
+		Short:            "A build automation tool for Docker applications",
+		SilenceUsage:     true,
+		SilenceErrors:    true,
+		TraverseChildren: true,
+		Args:             cobra.ArbitraryArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			opts.tasks = args
 			return runDobi(opts)

--- a/dockerfiles/Dockerfile.build
+++ b/dockerfiles/Dockerfile.build
@@ -1,8 +1,8 @@
 
-FROM    golang:1.6-alpine
+FROM    golang:1.7-alpine
 
 RUN     apk add -U git bash curl tree
-RUN     export GLIDE=v0.11.1; \
+RUN     export GLIDE=v0.12.3; \
         export SRC=https://github.com/Masterminds/glide/releases/download/; \
         curl -sL ${SRC}/${GLIDE}/glide-${GLIDE}-linux-amd64.tar.gz | \
         tar -xz linux-amd64/glide && \

--- a/dockerfiles/Dockerfile.docs
+++ b/dockerfiles/Dockerfile.docs
@@ -6,7 +6,9 @@ RUN     apk -U add \
             py-pip \
             go \
             bash \
-            git
+            git \
+            gcc \
+            musl-dev
 
 ENV     GOPATH=/go
 RUN     go get github.com/dnephin/filewatcher && \

--- a/glide.lock
+++ b/glide.lock
@@ -1,43 +1,42 @@
-hash: 8fe76a90b5508c1110299287bd59a2f16e33b4faab60013d464b5f49d0b6ad69
-updated: 2016-10-13T23:02:55.221904494Z
+hash: c4c91fb0ac9f49fa8e3745fdca0454db954c0c11eb638c04bf8fe5d32b64b6da
+updated: 2016-10-14T04:40:38.06676075Z
 imports:
 - name: github.com/Azure/go-ansiterm
   version: fa152c58bc15761d0200cb75fe958b89a9d4888e
   subpackages:
   - winterm
 - name: github.com/docker/docker
-  version: 487931902c1177352e4eceec1b5ef558a5ba24cc
+  version: 429812f86a7095c19ae48552f6b97150846ca29b
   subpackages:
-  - pkg/jsonmessage
-  - pkg/term
+  - api/types/filters
+  - api/types/mount
+  - api/types/swarm
+  - api/types/versions
   - opts
   - pkg/archive
   - pkg/fileutils
   - pkg/homedir
-  - pkg/stdcopy
-  - pkg/jsonlog
-  - pkg/system
-  - pkg/term/windows
   - pkg/idtools
   - pkg/ioutils
+  - pkg/jsonlog
+  - pkg/jsonmessage
+  - pkg/longpath
   - pkg/pools
   - pkg/promise
-  - pkg/longpath
-- name: github.com/docker/engine-api
-  version: f90ecdb1e989f834dabbd91807e891094aa069fe
-  subpackages:
-  - types/filters
-  - types/versions
+  - pkg/stdcopy
+  - pkg/system
+  - pkg/term
+  - pkg/term/windows
 - name: github.com/docker/go-connections
   version: 988efe982fdecb46f01d53465878ff1f2ff411ce
   subpackages:
   - nat
 - name: github.com/docker/go-units
-  version: f2d77a61e3c169b43402a0a1e84f06daf29b8190
+  version: f2145db703495b2e525c59662db69a7344b00bb8
 - name: github.com/fsouza/go-dockerclient
-  version: 107dc3807d24e581e874212dd9a7ba35599b893d
+  version: 5e508da6ac33603fcc7dc200b670ba295f24d975
 - name: github.com/gogits/git-module
-  version: 1dad9765c0adf04fe1eabadfb1a9189157dc50af
+  version: b3009dc4f5842cf9e2e80fef1e125e79c38e4949
 - name: github.com/golang/mock
   version: bd3c8e81be01eef76d4b503f5e687d2d1354d2d9
   subpackages:
@@ -52,40 +51,46 @@ imports:
   version: d52711f8d6bea8dc01efafdb68ad95a4e2606630
 - name: github.com/metakeule/fmtdate
   version: 427373e7d2f8fcaa70a83e289c15fb3618a945bc
+- name: github.com/Microsoft/go-winio
+  version: ce2922f643c8fd76b46cadc7f404a06282678b34
 - name: github.com/opencontainers/runc
-  version: 9b3a2f16345aa015ca286835f6f5eb67ea60aa4b
+  version: d186a7552b91af1270c2df74186fda6db2871869
   subpackages:
+  - libcontainer/system
   - libcontainer/user
 - name: github.com/renstrom/dedent
   version: 020d11c3b9c0c7a3c2efcc8e5cf5b9ef7bcea21f
 - name: github.com/Sirupsen/logrus
-  version: f3cfb454f4c209e6668c95216c4744b8fddb2356
+  version: 26709e2714106fb8ad40b773b711ebce25b78914
 - name: github.com/spf13/cobra
-  version: ec02ead39fbb63b2280d612f8ec0a4029c5a8e70
+  version: 8e4c18bd418c1b9060ddf7760316aa7c2a468127
   repo: https://github.com/dnephin/cobra
 - name: github.com/spf13/pflag
-  version: 367864438f1b1a3c7db4da06a2f55b144e6784e0
+  version: bf8481a6aebc13a8aab52e699ffe2e79771f5a3f
 - name: github.com/stretchr/testify
-  version: d77da356e56a7428ad25149ca77381849a6a5232
+  version: 976c720a22c8eb4eb6a0b4348ad85ad12491a506
   subpackages:
   - assert
-  - suite
   - require
+  - suite
+- name: github.com/Unknwon/com
+  version: 28b053d5a2923b87ce8c5a08f3af779894a72758
 - name: github.com/valyala/fasttemplate
   version: 3b874956e03f1636d171bda64b130f9135f42cff
 - name: golang.org/x/net
-  version: bc3663df0ac92f928d419e31e0d2af22e683a5a2
+  version: 8b4af36cd21a1f85a7484b49feb7c79363106d8e
   subpackages:
   - context
+  - context/ctxhttp
 - name: golang.org/x/sys
-  version: 62bee037599929a6e9146f29d10dd5208c43507d
+  version: 9bb9f0998d48b31547d975974935ae9b48c7a03c
   subpackages:
-  - unix
+  - windows
 - name: gopkg.in/yaml.v2
   version: a83829b6f1293c91addabc89d0571c246397bbf4
 testImports:
 - name: github.com/davecgh/go-spew
-  version: 5215b55f46b2b919f50a1df0eaa5886afe4e3b3d
+  version: 6d212800a42e8ab5c146b8ace3490ee17e5225f9
   subpackages:
   - spew
 - name: github.com/pmezard/go-difflib

--- a/glide.yaml
+++ b/glide.yaml
@@ -11,7 +11,6 @@ import:
   repo: https://github.com/dnephin/cobra
 - package: github.com/valyala/fasttemplate
 - package: github.com/gogits/git-module
-- package: github.com/mcuadros/go-version
 - package: github.com/metakeule/fmtdate
 - package: github.com/golang/mock
   subpackages:
@@ -19,3 +18,8 @@ import:
 - package: github.com/docker/go-connections
   subpackages:
   - nat
+- package: github.com/docker/docker
+  subpackages:
+  - pkg/jsonmessage
+  - pkg/term
+  - opts

--- a/glide.yaml
+++ b/glide.yaml
@@ -22,4 +22,3 @@ import:
   subpackages:
   - pkg/jsonmessage
   - pkg/term
-  - opts

--- a/script/watch
+++ b/script/watch
@@ -3,6 +3,7 @@
 exec filewatcher \
     -x '.git' \
     -x '.dobi' \
+    -x '.glide' \
     -x 'vendor' \
     -x '**/*.swp' \
     -x 'dist' \

--- a/tasks/image/pull.go
+++ b/tasks/image/pull.go
@@ -43,7 +43,7 @@ func now() *time.Time {
 }
 
 func pullImage(ctx *context.ExecuteContext, t *Task, imageTag string) error {
-	registry, err := parseRepo(t.config.Image)
+	registry, err := parseAuthRepo(t.config.Image)
 	if err != nil {
 		return err
 	}

--- a/tasks/image/push.go
+++ b/tasks/image/push.go
@@ -22,7 +22,7 @@ func RunPush(ctx *context.ExecuteContext, t *Task) error {
 }
 
 func pushImage(ctx *context.ExecuteContext, t *Task, tag string) error {
-	repo, err := parseRepo(t.config.Image)
+	repo, err := parseAuthRepo(t.config.Image)
 	if err != nil {
 		return err
 	}

--- a/tasks/image/utils.go
+++ b/tasks/image/utils.go
@@ -2,7 +2,6 @@ package image
 
 import (
 	"fmt"
-	"net/url"
 	"strings"
 
 	"github.com/dnephin/dobi/config"
@@ -33,24 +32,17 @@ func GetCanonicalTag(ctx *context.ExecuteContext, conf *config.ImageConfig) stri
 	return ctx.Env.Unique()
 }
 
-func parseRepo(image string) (string, error) {
-	repo, _ := docker.ParseRepositoryTag(image)
-	if !strings.HasPrefix(repo, "http") {
-		repo = "https://" + repo
-	}
-
-	addr, err := url.Parse(repo)
-	if err != nil {
-		return "", fmt.Errorf("Failed to parse repo name from %q: %s", repo, err)
-	}
-
-	if addr.Host == "" {
+func parseAuthRepo(image string) (string, error) {
+	// This is the approximate logic from
+	// github.com/docker/docker/reference.splitHostname(). That package is
+	// conflicting with other dependencies, so it can't be imported at this time.
+	parts := strings.SplitN(image, "/", 3)
+	switch len(parts) {
+	case 1, 2:
 		logging.Log.Debugf("Using default registry %q", defaultRepo)
 		return defaultRepo, nil
+	default:
+		logging.Log.Debugf("Using registry %q", parts[0])
+		return parts[0], nil
 	}
-
-	// TODO: what about v2?
-	addr.Path = "v1"
-	logging.Log.Debugf("Using registry %q", addr)
-	return addr.String(), nil
 }

--- a/tasks/image/utils_test.go
+++ b/tasks/image/utils_test.go
@@ -1,0 +1,25 @@
+package image
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseAuthRepoWithUserRepo(t *testing.T) {
+	repo, err := parseAuthRepo("dnephin/foo")
+	assert.Nil(t, err)
+	assert.Equal(t, repo, defaultRepo)
+}
+
+func TestParseAuthRepoPrivateRepoAndPort(t *testing.T) {
+	repo, err := parseAuthRepo("myrepo.net:3434/dnephin/foo")
+	assert.Nil(t, err)
+	assert.Equal(t, repo, "myrepo.net:3434")
+}
+
+func TestParseAuthRepoPrivateRepo(t *testing.T) {
+	repo, err := parseAuthRepo("myrepo.net/dnephin/foo")
+	assert.Nil(t, err)
+	assert.Equal(t, repo, "myrepo.net")
+}

--- a/tasks/job/run.go
+++ b/tasks/job/run.go
@@ -19,11 +19,13 @@ import (
 	"github.com/dnephin/dobi/tasks/image"
 	"github.com/dnephin/dobi/tasks/mount"
 	"github.com/dnephin/dobi/utils/fs"
-	dopts "github.com/docker/docker/opts"
 	"github.com/docker/docker/pkg/term"
 	"github.com/docker/go-connections/nat"
 	docker "github.com/fsouza/go-dockerclient"
 )
+
+// DefaultUnixSocket to connect to the docker API
+const DefaultUnixSocket = "/var/run/docker.sock"
 
 // Task is a task which runs a command in a container to produce a
 // file or set of files.
@@ -257,7 +259,7 @@ func provideDocker(opts docker.CreateContainerOptions) docker.CreateContainerOpt
 	case dockerHostEnv != "":
 		opts.Config.Env = append(opts.Config.Env, "DOCKER_HOST="+dockerHostEnv)
 	default:
-		path := dopts.DefaultUnixSocket
+		path := DefaultUnixSocket
 		opts.HostConfig.Binds = append(opts.HostConfig.Binds, path+":"+path)
 	}
 	return opts

